### ISSUE TITLE
Cannot Cast Counter To Double

### ISSUE THREAD

**Issue by jonny7** [scylladb/scylladb#14501]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [x] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash): 5.2.2
Cluster size: 3
OS (RHEL/CentOS/Ubuntu/AWS AMI):

*Hardware details (for performance issues)*          Delete if unneeded
Platform (physical/VM/cloud instance type/docker): Docker / Kubernetes

Trying to `CAST` a Counter to a double fails with 

> InvalidRequest: Error from server: code=2200 [Invalid query] message="org.apache.cassandra.db.marshal.CounterColumnType cannot be cast to org.apache.cassandra.db.marshal.DoubleType"
 
I've replicated this in Docker with version 5.1.7 + 5.2.2.

I've just tested this with Cassandra's latest Docker image too and it returns fine.

Thanks

**Comment by cvybhu** [comment]
Thank you for the bug report.
It looks like there's no implementation for converting `counter` values to `double`, we need to implement this feature.

A switch case needs to be added here:
https://github.com/scylladb/scylladb/blob/7d35cf8657ab593eebee31c5ae571c43d4a413a0/cql3/functions/castas_fcts.cc#L269-L283

As a temporary workaround, could you convert the value to `double` on the driver side?

**Comment by jonny7** [comment]
> Thank you for the bug report. It looks like there's no implementation for converting `counter` values to `double`, we need to implement this feature.
> 
> A switch case needs to be added here:
> 
> https://github.com/scylladb/scylladb/blob/7d35cf8657ab593eebee31c5ae571c43d4a413a0/cql3/functions/castas_fcts.cc#L269-L283
> 
> As a temporary workaround, could you convert the value to `double` on the driver side?

Thanks for this, it's Grafana that is doing the queries so in that case unfortunately not.

**Comment by nyh** [comment]
In `cql3/functions/castas_fcts.cc` we have a four-year-old comment:
```
// FIXME: Add conversions for counters, after they are fully implemented...
```
Maybe it's time to do that...

@mykaul @eliransin we also need to do some soul-searching why we didn't find this bug **much** earlier:
1. It appears the author of dtest's `cql_cast_test.py` was aware of this missing implementation and left behind a `TODO: counters are not supported. should be added later` instead of creating xfailing tests.
2. The author of `test/cql-pytest/test_cast_data.py` (that's me...) created this file to check one specific bug, and was too much in a hurry to write tests for all types and left a `# TODO: test casts from more types.`.
3. Cassandra has a unit test for these casts, which would have caught this bug - `functions/CastFctsTest.java`. If the author of the test translation project #7722 (again, that's me...) put more effort into it, we would have known about this bug already.

Of course, now that we do know about this bug, we also need to fix it.

**Comment by nyh** [comment]
I'm removing the "enhancement" tag - as the user noted, missing small details in CQL can break existing applications, and will be perceived by users as a bug, not a request for a new feature.

**Comment by mykaul** [comment]
Maybe they aren't a popular use case. I wonder what Grafana input is.

**Comment by nyh** [comment]
@mykaul neither counters not casts are very "popular" features, so their combination is probably even less popular - but they do exist, and Grafana (which I didn't even know could use Cassandra as a backend, nice!) might not be the only Cassandra application that uses them.

**Comment by nyh** [comment]
I could easily make the Cassandra test for counters work, but I think I opened Pandora's box :-( Even after I get `CAST(c AS double)` et al. to work, the silly `CAST(c as counter)` doesn't work - it invokes one of several `fail(unimplemented::cause::COUNTERS)` in type.cc. 

So now I think that all these `unimplemented::cause::COUNTERS` are bugs - we do have counters implemented so we shouldn't have any of those "unimplemented" things left.

Recently in b25ee62f75268981a8606d8bede2c73482a25612 @wmitros also needed to deserialize counters and had to use weird workarounds because of these "unimplemented" things.

I'm actually really worried that some of this code that reads counter, including the one that @wmitros wrote and I'm writing now, might be wrong. The code assumes that a counter is just a 64-bit integer. But isn't counter actually a complicated CRDT object, and needs to be "evaluated" to figure out the real count? I'm worried that my test (which increments a counter in memtable, so it might be short-circuited?) doesn't actually exercise this case. Or, maybe it's fine to read a counter as a 64-bit integer because we already have some other code which implicitly does the calculation when a counter object is converted to a 64-bit integer?

**Comment by cvybhu** [comment]
I ran into the same thing. I thought it's just a quick fix, but the `counter_type_impl` isn't fully implemented :C
The people who originally introduced the counter type didn't implement all the methods for this type.

They have to be implemented, but this require much more work.
For example, can `counter` values be `empty`?

**Comment by cvybhu** [comment]
The good news is that once I implemented all the methods by calling the `long_type_impl` ones, the test started working. I could cast something to counter and back. Still I'm not sure how correct this is due to empty values.

**Comment by nyh** [comment]
There's something basic I don't understand. On-disk, a counter is some complicated thing with "shards" from different sstables, and so on. Am I correct in understanding that the representation in memory is always the sum of everything as just a uint64_type? If that's true,  why isn't `counter_type` simply an integer_type() like long_type()? Why is `counter_type` a subclass of `abstract_type` instead? Was there a reason why it had to be complicated this way?

@avikivity @mykaul who is the expert in the counters column type in the company?

**Comment by denesb** [comment]
> Am I correct in understanding that the representation in memory is always the sum of everything as just a uint64_type?

No, counters are also represented as shards in memory too. But counters are conceptually a 64 bit integer, to calculate their value, you simply sum up all the shards.

We also have another type of (legacy?) counter, that I found traces of in the code. I have no idea what that is, maybe that is the remains of an older implementation.

> @avikivity @mykaul who is the expert in the counters column type in the company?

Sadly, all counter experts have left the company. So we are left with reverse-engineering the code. Some people (Avi) might remember why things are they way they are.

I was always irked by how the counters were different then all the other types, with most regular type methods being unimplemented. My best guess is that this is simply due to we never getting around to properly finish this feature. I think it is time to remove all those nasty "unimplemented" traps.

**Comment by avikivity** [comment]
Counters are just bigints with a funny name on the coordinator side (there's a back door via CQL SCYLLA_COUNTER_SHARD_LIST) . They're only special on the replica.

**Comment by avikivity** [comment]
> > Counters are just bigints with a funny name on the coordinator side (there's a back door via CQL SCYLLA_COUNTER_SHARD_LIST) . They're only special on the replica.
> 
> So Avi, is there a reason why counter_type cannot be an `integer_type()` like `long_type` and inherit all the types.cc functions, instead of being an unimplemented special case everywhere? 

The actual type in the schema has to be different, otherwise we don't know it's a counter.

> However, I guess it will be simpler for me to just implement those special cases using the long_type implementation (just like @cvybhu said) instead of trying to think what I might ruin in other places in the code by changing counter_type's defintion...

Yes.

**Comment by nyh** [comment]
> > They have to be implemented, but this requires much more work. For example, can `counter` values be `empty`?
> 
> Anything can be empty through ugly tricks (see `test_empty.py::test_empty_int` for example of empty int)

On second thought, I don't think a real counter can be "empty". Counters are written differently, and I don't see how an "empty" value can be written there. I wonder what happens, though, if you cast an "empty integer" (silly, but existing, concept) to other types, including counters. I guess yet another thing to test...

**Comment by nyh** [comment]
I already have a working implementation that works well (and not too ugly) and will send a PR soon. 

As usual, as G. I. Joe said, "knowing is half the battle" - tests (which are the vast majority of the content of my patch) exposed weird corner cases and even more unimplemented functions I didn't find in the first try.

**Comment by nyh** [comment]
> On second thought, I don't think a real counter can be "empty". Counters are written differently, and I don't see how an "empty" value can be written there. I wonder what happens, though, if you cast an "empty integer" (silly, but existing, concept) to other types, including counters. I guess yet another thing to test...

It turns out (I wrote a test for this as well...) that neither Cassandra nor Scylla allow casting other types to "counter", so you can't cast an empty integer to a counter. Moreover, because of various limitations of CQL, I couldn't check what `blobAsCounter(0x)` (this is the trick often used to create an "empty integer") actually does, so I don't really know if it works as expected or not - there is no statement I could think of where this construct can actually be used.

**Comment by jonny7** [comment]
Any idea @nyh what release/when this fix will be released?

**Comment by nyh** [comment]
@jonny7 it is already available in the nightly image so you can try that out right away. But it's not in any released version: It will be in the next open-source version but there is not schedule for when this will happen (it hasn't been branched yet).

As I noted in https://github.com/scylladb/scylladb/issues/14501#issuecomment-1620481883, I think we should consider this a compatibility bug, not a feature, so I want to backport it to the 5.2 and 5.1 branches. When I do that (stay tuned), you'll need to wait for the next 5.2.* or 5.1.* point releases to get this fix.

**Comment by nyh** [comment]
Backported to 5.2:   
The backport of a637ddd09c3596d6838d3d562f713b7ee5f00a34 was clean only after also backporting three test-only patches that created the tests and test functions needed for the tests included in the backported patch: 78555ba7f14cb5517637aaabf0decae97ac21547, f08bc83cb2beed919269fb759efc2caf693a7597,  328cdb2124aa3004893f9b8096f85157091b0985.
The backported version includes the tests, and I verified that they (and all cql-pytest in general) passes.

**Comment by nyh** [comment]
Backported to 5.1:
Again required backporting the same three test patches and yet another test patch 0c26032e705cee98409113a5a3c07652efc82939 for more missing test functions, and finally  https://github.com/scylladb/scylladb/commit/a637ddd09c3596d6838d3d562f713b7ee5f00a34 (there was one small modification I needed to do in one of the function calls in that patch where the API changed a big).
Unfortunately, one of the backported tests  - an irrelevant test that "unset" values in a request produce an error - doesn't pass on this old version, so I manually commented it out. This test  it's not relevant to this issue (casting counters) and it's just an error handling case, not a bugfix we would backport anyway so I don't fill compelled to figure out which patch fixed that bug and backport it too.

Again I verified manually that all cql-pytest tests pass after the backport.

After backporting to both 5.2 and 5.1, I will remove the "backport candidate" tag.
